sql/logictest: TestLogic failed

[(sql/logictest).TestLogic failed](https://viewLog.html?buildId=1655563&tab=buildLog) on [master@47de313ba8c3026476ab928a3fa19f2c5b972191](https://github.com/cockroachdb/cockroach/commits/47de313ba8c3026476ab928a3fa19f2c5b972191):

```
=== RUN   TestLogic
--- FAIL: TestLogic (638.73s)
    test_log_scope.go:77: test logs captured to: /tmp/logTestLogic706898202
    test_log_scope.go:58: use -show-logs to present logs inline
=== RUN   TestLogic/fakedist-vec-off
    --- FAIL: TestLogic/fakedist-vec-off (0.82s)
=== RUN   TestLogic/fakedist-vec-off/schema_change_in_txn/delete_index_and_table_in_txn
rewrote:
CREATE TABLE x (a INT8 PRIMARY KEY, b INT8, c INT8 AS (b + 1) STORED, d INT8 AS (b - 1) STORED, FAMILY (c), FAMILY (d), FAMILY (b));


            --- FAIL: TestLogic/fakedist-vec-off/schema_change_in_txn/delete_index_and_table_in_txn (0.42s)
                logic.go:1891: 
                    
                    testdata/logic_test/schema_change_in_txn:1742: 
                    expected success, but found
                    (40001) restart transaction: TransactionRetryWithProtoRefreshError: TransactionRetryError: retry txn (RETRY_SERIALIZABLE): "sql txn" meta={id=6d0062d3 key=/Table/SystemConfigSpan/Start pri=0.01281095 epo=0 ts=1576774383.185245872,1 min=1576774383.163685859,0 seq=29} rw=true stat=PENDING rts=1576774383.163685859,0 wto=false max=1576774383.163685859,0
                logic.go:1304: 
                     pq: restart transaction: TransactionRetryWithProtoRefreshError: TransactionRetryError: retry txn (RETRY_SERIALIZABLE): "sql txn" meta={id=6d0062d3 key=/Table/SystemConfigSpan/Start pri=0.01281095 epo=0 ts=1576774383.185245872,1 min=1576774383.163685859,0 seq=29} rw=true stat=PENDING rts=1576774383.163685859,0 wto=false max=1576774383.163685859,0
=== RUN   TestLogic/fakedist-vec-off/schema_change_in_txn
=== PAUSE TestLogic/fakedist-vec-off/schema_change_in_txn
=== CONT  TestLogic/fakedist-vec-off/schema_change_in_txn
        --- FAIL: TestLogic/fakedist-vec-off/schema_change_in_txn (31.37s)
```

<details><summary>Repro</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestLogic PKG=./pkg/sql/logictest TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```

<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql/logictest).TestLogic failed](https://viewLog.html?buildId=1660202&tab=buildLog) on [master@319080c701cb4e13c347e86c4afb2f1c2af78def](https://github.com/cockroachdb/cockroach/commits/319080c701cb4e13c347e86c4afb2f1c2af78def):

```
=== RUN   TestLogic
--- FAIL: TestLogic (729.65s)
    test_log_scope.go:77: test logs captured to: /tmp/logTestLogic630471527
    test_log_scope.go:58: use -show-logs to present logs inline
=== RUN   TestLogic/local/schema_change_in_txn/delete_index_and_table_in_txn
rewrote:
CREATE TABLE c2 (id STRING PRIMARY KEY, b_id STRING, x INT8, FOREIGN KEY (b_id, x) REFERENCES b1 (id, x) MATCH FULL ON UPDATE CASCADE, FAMILY (b_id), FAMILY (x));


            --- FAIL: TestLogic/local/schema_change_in_txn/delete_index_and_table_in_txn (0.73s)
                logic.go:1891: 
                    
                    testdata/logic_test/schema_change_in_txn:1742: 
                    expected success, but found
                    (40001) restart transaction: TransactionRetryWithProtoRefreshError: TransactionRetryError: retry txn (RETRY_SERIALIZABLE): "sql txn" meta={id=56f88e55 key=/Table/SystemConfigSpan/Start pri=0.01255604 epo=0 ts=1576944055.134075282,1 min=1576944055.095443779,0 seq=29} rw=true stat=PENDING rts=1576944055.095443779,0 wto=false max=1576944055.095443779,0
                logic.go:1304: 
                     pq: restart transaction: TransactionRetryWithProtoRefreshError: TransactionRetryError: retry txn (RETRY_SERIALIZABLE): "sql txn" meta={id=56f88e55 key=/Table/SystemConfigSpan/Start pri=0.01255604 epo=0 ts=1576944055.134075282,1 min=1576944055.095443779,0 seq=29} rw=true stat=PENDING rts=1576944055.095443779,0 wto=false max=1576944055.095443779,0
=== RUN   TestLogic/local/schema_change_in_txn
=== PAUSE TestLogic/local/schema_change_in_txn
=== CONT  TestLogic/local/schema_change_in_txn
rewrote:
CREATE TABLE test.xyz (x INT8 PRIMARY KEY, y INT8, z INT8, INDEX foo (z, y), FAMILY (y), FAMILY (z));


rewrote:
CREATE TABLE test.xz (y INT8, z INT8, INDEX foo (z, y), FAMILY (z, y));


--- done: testdata/logic_test/postgresjoin with config local: 101 tests, 0 failures
        --- FAIL: TestLogic/local/schema_change_in_txn (30.63s)
=== RUN   TestLogic/local
    --- FAIL: TestLogic/local (0.47s)
```

<details><summary>Repro</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestLogic PKG=./pkg/sql/logictest TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```

<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql/logictest).TestLogic failed](https://viewLog.html?buildId=1693004&tab=buildLog) on [master@e7f83bf05d032e374e80f03c2355fa855d62e285](https://github.com/cockroachdb/cockroach/commits/e7f83bf05d032e374e80f03c2355fa855d62e285):

```
--- progress: testdata/logic_test/fk_opt: 742 statements/queries
--- done: testdata/logic_test/vectorize_expr with config local: 6 tests, 0 failures
rewrote:
CREATE TABLE t (a INT8 PRIMARY KEY, b INT8, FAMILY (a), FAMILY (b));


        --- FAIL: TestLogic/local/views (5.91s)
            logic.go:1891: 
                
                testdata/logic_test/views:560: SELECT * FROM v1
                expected success, but found
                (XX000) internal error: unexpected error from the vectorized runtime: interface conversion: coldata.column is []bool, not []int64
                error.go:70: in func1()
                DETAIL: stack trace:
                github.com/cockroachdb/cockroach/pkg/sql/colexec/execerror/error.go:70: func1()
                runtime/panic.go:679: gopanic()
                runtime/iface.go:255: panicdottypeE()
                github.com/cockroachdb/cockroach/pkg/col/coldata/vec.go:202: Int64()
                github.com/cockroachdb/cockroach/pkg/sql/colexec/mergejoiner_leftouter.eg.go:30062: func1()
                github.com/cockroachdb/cockroach/pkg/sql/colexec/allocator.go:92: PerformOperation()
                github.com/cockroachdb/cockroach/pkg/sql/colexec/mergejoiner_leftouter.eg.go:28391: buildLeftGroups()
                github.com/cockroachdb/cockroach/pkg/sql/colexec/mergejoiner_leftouter.eg.go:32522: build()
                github.com/cockroachdb/cockroach/pkg/sql/colexec/mergejoiner_leftouter.eg.go:32604: Next()
                github.com/cockroachdb/cockroach/pkg/sql/colexec/invariants_checker.go:45: Next()
                github.com/cockroachdb/cockroach/pkg/sql/colexec/fn_op.go:35: Next()
                github.com/cockroachdb/cockroach/pkg/sql/colexec/distinct.eg.go:2002: Next()
                github.com/cockroachdb/cockroach/pkg/sql/colexec/aggregator.go:354: Next()
                github.com/cockroachdb/cockroach/pkg/sql/colexec/is_null_ops.go:50: Next()
                github.com/cockroachdb/cockroach/pkg/sql/colexec/simple_project.go:93: Next()
                github.com/cockroachdb/cockroach/pkg/sql/colexec/invariants_checker.go:45: Next()
                github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:149: next()
                github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:140: nextAdapter()
                github.com/cockroachdb/cockroach/pkg/sql/colexec/execerror/error.go:91: CatchVectorizedRuntimeError()
                github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:177: Next()
                github.com/cockroachdb/cockroach/pkg/sql/execinfra/base.go:170: Run()
                github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:749: Run()
                github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:370: Run()
                github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:375: Run()
                github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:977: PlanAndRun()
                github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:823: execWithDistSQLEngine()
                github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:715: dispatchToExecutionEngine()
                github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:418: execStmtInOpenState()
                github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:100: execStmt()
                github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1283: execCmd()
                github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1212: run()
                github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:451: ServeConn()
                
                NOTE: internal errors may have more details in logs. Use -show-logs.
            logic.go:1652: 
                 pq: internal error: unexpected error from the vectorized runtime: interface conversion: coldata.column is []bool, not []int64
```

<details><summary>Repro</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestLogic PKG=./pkg/sql/logictest TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```

<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql/logictest).TestLogic failed](https://viewLog.html?buildId=1697835&tab=buildLog) on [master@13af75f3d1376fff31d1419df2951c2d8a810efd](https://github.com/cockroachdb/cockroach/commits/13af75f3d1376fff31d1419df2951c2d8a810efd):

```

        --- FAIL: TestLogic/fakedist-metadata/distsql_automatic_stats (53.15s)
            logic.go:1898: 
                
                testdata/logic_test/distsql_automatic_stats:58: SELECT statistics_name, column_names, row_count, distinct_count, null_count
                FROM [SHOW STATISTICS FOR TABLE data] ORDER BY column_names::STRING, created
                expected success, but found
                (XXUUU) read-table-stats: expected 3 metadata senders but found 0; missing 66260a3d-ffca-403c-887e-ee91f2bece44, eea63b39-b5b1-4091-b277-126207a299b9, dae258c0-fc98-445b-84ef-57bb70615db3
                internal.go:477: in func1()
            logic.go:1640: condition failed to evaluate within 45s: 
                testdata/logic_test/distsql_automatic_stats:58: SELECT statistics_name, column_names, row_count, distinct_count, null_count
                FROM [SHOW STATISTICS FOR TABLE data] ORDER BY column_names::STRING, created
                expected:
                    statistics_name  column_names  row_count  distinct_count  null_count
                    __auto__         {a}           1000       10              0
                    __auto__         {a}           1000       10              0
                    __auto__         {b}           1000       10              0
                    __auto__         {b}           1000       10              0
                    __auto__         {c}           1000       10              0
                    __auto__         {c}           1000       10              0
                    __auto__         {d}           1000       1               1000
                    __auto__         {d}           1000       2               730
                    
                but found (query options: "colnames,rowsort,retry" -> ignore the following ordering of rows) :
                    statistics_name  column_names  row_count  distinct_count  null_count
                    __auto__         {a}           1000       10              0
                    __auto__         {b}           1000       10              0
                    __auto__         {c}           1000       10              0
                    __auto__         {d}           1000       1               1000
                    
                
                goroutine 3790168 [running]:
                runtime/debug.Stack(0xc00b2d3658, 0x403f640, 0xc042f31c20)
                	/usr/local/go/src/runtime/debug/stack.go:24 +0x9d
                github.com/cockroachdb/cockroach/pkg/testutils.SucceedsSoon(0x4120e20, 0xc00893ef00, 0xc00b2d3658)
                	/go/src/github.com/cockroachdb/cockroach/pkg/testutils/soon.go:37 +0x6b
                github.com/cockroachdb/cockroach/pkg/sql/logictest.(*logicTest).processSubtest(0xc018a4a180, 0x0, 0x0, 0xc0361b9c50, 0x0, 0xc0005d9410, 0x2b, 0x3947c30, 0x11, 0x3, ...)
                	/go/src/github.com/cockroachdb/cockroach/pkg/sql/logictest/logic.go:1640 +0x2055
                github.com/cockroachdb/cockroach/pkg/sql/logictest.(*logicTest).processTestFile(0xc018a4a180, 0xc0005d9410, 0x2b, 0x3947c30, 0x11, 0x3, 0x1, 0x392a267, 0x2, 0x0, ...)
                	/go/src/github.com/cockroachdb/cockroach/pkg/sql/logictest/logic.go:1293 +0x699
                github.com/cockroachdb/cockroach/pkg/sql/logictest.(*logicTest).runFile(0xc018a4a180, 0xc0005d9410, 0x2b, 0x3947c30, 0x11, 0x3, 0x1, 0x392a267, 0x2, 0x0, ...)
                	/go/src/github.com/cockroachdb/cockroach/pkg/sql/logictest/logic.go:2204 +0x110
                github.com/cockroachdb/cockroach/pkg/sql/logictest.RunLogicTest.func1.1(0xc00893ef00)
                	/go/src/github.com/cockroachdb/cockroach/pkg/sql/logictest/logic.go:2328 +0x266
                testing.tRunner(0xc00893ef00, 0xc02b795260)
                	/usr/local/go/src/testing/testing.go:909 +0xc9
                created by testing.(*T).Run
                	/usr/local/go/src/testing/testing.go:960 +0x350
=== RUN   TestLogic/fakedist-metadata
    --- FAIL: TestLogic/fakedist-metadata (0.53s)

```

<details><summary>Repro</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestLogic PKG=./pkg/sql/logictest TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```

[roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestLogic.%2A&sort=title&restgroup=false&display=lastcommented+project)

<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/colexec/buffered_batch.go
+++ b/pkg/sql/colexec/buffered_batch.go
@@ -29,6 +29,9 @@ func newBufferedBatch(allocator *Allocator, types []coltypes.T, initialSize int)
 // bufferedBatch is a custom implementation of coldata.Batch interface (only
 // a subset of methods is implemented) which stores the length as uint64. It
 // should be used whenever we're doing possibly unbounded buffering of tuples.
+// WARNING: do *not* use Length() method on bufferedBatch to get the number of
+// buffered tuples because it might return incorrect result. Instead, use
+// 'length' field directly.
 type bufferedBatch struct {
 	colVecs []coldata.Vec
 	length  uint64
@@ -36,11 +39,13 @@ type bufferedBatch struct {
 
 var _ coldata.Batch = &bufferedBatch{}
 
+// Length method should *only* be used to check whether the batch has a zero
+// length.
 func (b *bufferedBatch) Length() uint16 {
-	execerror.VectorizedInternalPanic("Length() should not be called on bufferedBatch; instead, " +
-		"length field should be accessed directly")
-	// This code is unreachable, but the compiler cannot infer that.
-	return 0
+	if b.length == 0 {
+		return 0
+	}
+	return 1
 }
 
 func (b *bufferedBatch) SetLength(uint16) {

--- a/pkg/sql/colexec/mergejoiner_tmpl.go
+++ b/pkg/sql/colexec/mergejoiner_tmpl.go
@@ -752,7 +752,7 @@ func (o *mergeJoin_JOIN_TYPE_STRING_FILTER_INFO_STRINGOp) buildLeftGroups(
 				outStartIdx := int(destStartIdx)
 				out := o.output.ColVec(outColIdx)
 				var src coldata.Vec
-				if batch.Width() > int(inColIdx) {
+				if batch.Length() > 0 {
 					src = batch.ColVec(int(inColIdx))
 				}
 				colType := input.sourceTypes[inColIdx]
@@ -930,7 +930,7 @@ func (o *mergeJoin_JOIN_TYPE_STRING_FILTER_INFO_STRINGOp) buildRightGroups(
 				outStartIdx := int(destStartIdx)
 				out := o.output.ColVec(outColIdx + colOffset)
 				var src coldata.Vec
-				if batch.Width() > int(inColIdx) {
+				if batch.Length() > 0 {
 					src = batch.ColVec(int(inColIdx))
 				}
 				colType := input.sourceTypes[inColIdx]

--- a/pkg/sql/logictest/logic.go
+++ b/pkg/sql/logictest/logic.go
@@ -535,8 +535,10 @@ var (
 	defaultConfigNames = []string{
 		"local",
 		"local-vec-off",
+		"local-vec",
 		"fakedist",
 		"fakedist-vec-off",
+		"fakedist-vec",
 		"fakedist-metadata",
 		"fakedist-disk",
 	}

--- a/pkg/sql/logictest/testdata/logic_test/exec_merge_join_dist
+++ b/pkg/sql/logictest/testdata/logic_test/exec_merge_join_dist
@@ -1,7 +1,5 @@
 # LogicTest: 5node-dist-vec
 
-skip #43357
-
 # Regression test for #39317.
 
 statement ok

--- a/pkg/sql/logictest/testdata/logic_test/scrub
+++ b/pkg/sql/logictest/testdata/logic_test/scrub
@@ -1,5 +1,3 @@
-skip #43357
-
 statement error pgcode 42P01 relation "t" does not exist
 EXPERIMENTAL SCRUB TABLE t
 -----

--- a/pkg/sql/logictest/testdata/logic_test/srfs
+++ b/pkg/sql/logictest/testdata/logic_test/srfs
@@ -204,7 +204,7 @@ INSERT INTO t VALUES ('cat')
 statement ok
 INSERT INTO u VALUES ('bird')
 
-query TTII colnames
+query TTII colnames,rowsort
 SELECT t.*, u.*, generate_series(1,2), generate_series(3, 4) FROM t, u
 ----
 a    b     generate_series generate_series

--- a/pkg/sql/logictest/testdata/logic_test/views
+++ b/pkg/sql/logictest/testdata/logic_test/views
@@ -1,5 +1,3 @@
-skip #43357
-
 # NOTE: Keep this table at the beginning of the file to ensure that its numeric
 #       reference is 53 (the numeric reference of the first table). If the
 #       numbering scheme in cockroach changes, this test will break.

